source/plugin/optional/basic.rst

```diff
@@ -0,0 +1,195 @@
+================================================
+
+Let's say we have a simple API object ``Entity`` with a ``getFoo()`` method which returns the Entity's ``Foo``.
+
+.. image:: /images/optionals1.png
+
+In the olden times of yore, our plugin might fetch and use the ``Foo`` from the entity using the ``getter`` like this:
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Foo entityFoo = someEntity.getFoo();
+        entityFoo.bar();
+    }
+
+The problem arises because - when designing the API - we have to rely an *implicit* contract on the ``getFoo`` method
+with respect to whether the method can (or cannot) return ``null``. This *implicit contract* can be defined in one of
+two ways:
+
+* **In the javadoc** - this is bad because it relies on the plugin author reading the javadoc for the method, and the contract may not be clear to the plugin author
+* **Using nullable annotations** - this is not ideal because in general these annotations require a tool to be of any use, for example relying on the IDE or compiler to handle the annotations.
+
+.. image:: /images/optionals2.png
+
+Let's assume that the ``getFoo()`` method can - as part of its contract - return null. This suddenly means that our
+code above is unsafe as it may result in a ``NullPointerException`` if ``entityFoo`` is null.
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Foo entityFoo = someEntity.getFoo();
+        entityFoo.bar();
+    }
+
+Let's assume our plugin author is savvy to the nullable nature of our ``getFoo`` method and decides to fix the problem
+with null checking. Assuming he has defined a local constant ``Foo``, he resultant code looks like this:
+
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Foo entityFoo = someEntity.getFoo();
+        if (entityFoo == null) {
+            entityFoo = MyPlugin.DEFAULT_FOO;
+        }
+        entityFoo.bar();
+    }
+
+In this example, the plugin author is aware that the method can return null and has a constant available with a
+default instance of ``Foo`` which can be used instead. Of course the plugin could just short-circuit the call entirely,
+or it could attempt to fetch `Foo` from somewhere else. The key take-away is that handling nulls even in simple cases
+can lead to spaghetti code quite quickly, and moreover relies on the plugin author to explicitly visit the method's
+contract to check whether a null check is necessary in the first place.
+
+That's not the only drawback however. Let's consider the API over the longer term and assume that at the time the author
+writes their plugin, they visit the method javadoc and see that the method is guaranteed to never return null
+(since every `Entity` always has a ``Foo`` available). Great! No convoluted null check required!
+
+However, let's now assume that several versions of the game along the line, the game developers remove or deprecate
```

Comment:
I presume "along the line" means in the future? This is poorly worded. Perhaps "in subsequent versions"?